src/SpendPermissionManager.sol

```diff
@@ -478,8 +499,18 @@ contract SpendPermissionManager is EIP712 {
 
     /// @notice Approve spend permission.
     ///
+    /// @dev Emits a {SpendPermissionApproved} event if the spend permission is newly approved and not already revoked.
+    ///
     /// @param spendPermission Details of the spend permission.
-    function _approve(SpendPermission memory spendPermission) internal {
+    ///
+    /// @return approved True if spend permission is approved and not revoked.
+    function _approve(SpendPermission memory spendPermission) internal returns (bool) {
+        // return false early if spend permission is already revoked
+        if (_isRevoked[getHash(spendPermission)]) return false;
+
+        // return early if spend permission is already approved
+        if (_isApproved[getHash(spendPermission)]) return true;
```

Comment:
Can we reuse a single `bytes32 hash = getHash(spendPermission)`? I think we have one at the bottom we can move above these and share.

Also, thoughts on moving all state-related operations to the bottom for logical grouping? Code would read as all static checks at the top, then these stateful checks, then state setting.